src/gei/Commands/GenerateArchiveCommand.cs

```diff
@@ -0,0 +1,150 @@
+
+            if (string.IsNullOrWhiteSpace(ghesUrl))
+            {
+                _log.LogInformation("ghesUrl not provided, defaulting to https://api.github.com");
+                ghesUrl = "https://api.github.com";
+            }
+
+            if (noSslVerify)
+            {
+                _log.LogInformation("No SSL verification enabled");
+            }
+
+            var githubApi = noSslVerify ? _sourceGithubApiFactory.CreateClientNoSSL() : _sourceGithubApiFactory.Create();
+
+            var repositories = new string[] { sourceRepo };
+
+            var gitDataOptions = new
+            {
+                repositories,
+                exclude_metadata = true
+            };
+            var gitDataArchiveId = await githubApi.StartArchiveGeneration(ghesUrl, githubSourceOrg, gitDataOptions);
+
+            _log.LogInformation($"Archive generation of git data started with id: {gitDataArchiveId}");
+
+            var metadataOptions = new
+            {
+                repositories,
+                exclude_git_data = true,
+                exclude_releases = true,
+                exclude_owner_projects = true
+            };
+            var metadataArchiveId = await githubApi.StartArchiveGeneration(ghesUrl, githubSourceOrg, metadataOptions);
+
+            _log.LogInformation($"Archive generation of metadata started with id: {metadataArchiveId}");
+
+
+            var ids = new int[] { gitDataArchiveId, metadataArchiveId };
+
+            var archiveIdsWithFinished = new Dictionary<int, bool>()
+            {
+                { gitDataArchiveId, false },
+                { metadataArchiveId, false }
+            };
+
+            var currFinished = 0;
+            var total = ids.Length;
+            var timeOut = DateTime.Now.AddHours(_timeoutInHours);
+
+            while (currFinished < total && DateTime.Now < timeOut)
+            {
+                foreach (var pair in archiveIdsWithFinished)
+                {
+                    if (pair.Value)
+                    {
+                        continue;
+                    }
+                    var id = pair.Key;
+                    var archiveStatus = await githubApi.GetArchiveMigrationStatus(ghesUrl, githubSourceOrg, id);
+                    var stringStatus = GithubEnums.ArchiveMigrationStatusToString(archiveStatus);
+
+                    _log.LogInformation($"Waiting for archive with id {id} generation to finish. Current status: {stringStatus}");
+
+                    if (archiveStatus == GithubEnums.ArchiveMigrationStatus.Exported)
+                    {
+                        archiveIdsWithFinished[id] = true;
+                        currFinished++;
+                    }
+                    else if (archiveStatus == GithubEnums.ArchiveMigrationStatus.Failed)
+                    {
+                        _log.LogError($"Archive generation failed with id: {id}");
+                        return;
+                    }
+
+                }
```

Comment:
This `foreach` logic can easily be converted to a method. It seems like all you care is whether the status is `Exported` or `Failed`. So the method can just return the enum. Then in your while loop instead of the `foreach` you can just call the method twice (once per id) and decide whether to continue or not instead of incrementing `currFinished` variable. Something like this:

```c#
var gitDataArchiveMigrationStatus = ArchiveMigrationStatus.Pending;
var metadataArchiveMigrationStatus = ArchiveMigrationStatus.Pending;
while (gitDataArchiveMigrationStatus != ArchiveMigrationStatus.Exported &&
       metadataArchiveMigrationStatus != ArchiveMigrationStatus.Exported &&  
       DateTime.Now < timeout)
{
  gitDataArchiveMigrationStatus = gitDataArchiveMigrationStatus || QueryArchiveMigrationStatus(gitDataArchiveId);
  if (gitDataArchiveMigrationStatus == ArchiveMigrationStatus.Failed)
  {
    return;
  }

  metadataArchiveMigrationStatus = metadataArchiveMigrationStatus || QueryArchiveMigrationStatus(metadataArchiveId);
  if (metadataArchiveMigrationStatus == ArchiveMigrationStatus.Failed)
  {
    return;
  }
}
```

IMO this is way more readable than using dictionary, etc. `C#` supports local functions too (functions inside functions), so you can also create one for checking the overall status, i.e. `ShouldContinue(...)` and use it in the while loop's predicate 😃

Reply:
Accepting the other comment refactor 👍